packages/typegpu/src/resolutionCtx.ts

```diff
@@ -645,7 +650,21 @@ export class ResolutionCtxImpl implements ResolutionCtx {
     }
 
     if (schema && isWgslArray(schema)) {
-      return `array(${
+      if (!Array.isArray(value)) {
+        throw new WgslTypeError(
+          `Cannot coerce ${value} into value of type '${schema}'`,
+        );
+      }
+      if (schema.elementCount !== (value as unknown[]).length) {
+        throw new WgslTypeError(
+          `Cannot create value of type '${schema}' from an array of length: ${
+            (value as unknown[]).length
+          }`,
+        );
+      }
```

Comment:
We asserted it earlier so TS knows
```suggestion
      if (schema.elementCount !== value.length) {
        throw new WgslTypeError(
          `Cannot create value of type '${schema}' from an array of length: ${
            value.length
          }`,
        );
      }
```